### ISSUE TITLE
Implement Phase 1 Step 1: Synthetic Data Generator Implementation

### ISSUE THREAD

**Issue by dhar174** [dhar174/steel_defect_demo#15]
# Synthetic Data Generator Implementation

## Overview
Implement the `SteelCastingDataGenerator` class as the foundation for generating realistic synthetic steel casting process data. This is Phase 1, Step 1 of the Technical Implementation Specification.

### File Structure
- **Primary File**: `src/data/data_generator.py`
- **Configuration**: `configs/data_generation.yaml`
- **Script**: `scripts/generate_synthetic_data.py`

### Key Dependencies
```python
import numpy as np
import pandas as pd
import yaml
from pathlib import Path
from typing import Dict, Tuple, List
import json
```

### 1. SteelCastingDataGenerator Class
Implement the main generator class with the following methods:

#### Constructor (`__init__`)
- Load configuration from YAML file
- Initialize random state with configured seed
- Set up sensor parameters and defect simulation rules

#### `generate_cast_sequence(cast_id: str) -> Tuple[pd.DataFrame, Dict]`
Generate time series data for a single cast with:
- **Duration**: 120 minutes per cast (configurable)
- **Sampling Rate**: 1 Hz (configurable)
- **Sensors**:
  - Casting Speed: base 1.2 m/min, noise std 0.05, range 0.8-1.8
  - Mold Temperature: base 1520°C, noise std 10, range 1480-1580
  - Mold Level: base 150mm, noise std 5, range 120-180
  - Cooling Water Flow: base 200 L/min, noise std 15, range 150-250
  - Superheat: base 25°C above liquidus, noise std 3, range 15-40
- **Timestamp Generation**: Proper datetime indexing
- **Metadata**: Cast ID, generation timestamp, steel grade info

#### `generate_dataset() -> None`
Generate complete dataset:
- **Total Casts**: 1200 (configurable)
- **Train/Test Split**: 80/20
- **Output Formats**: Parquet for time series, JSON for metadata
- **Directory Structure**: Save to `data/raw/`, `data/processed/`, `data/synthetic/`

### 2. Defect Simulation Logic
Implement realistic defect triggering based on:

#### Defect Triggers
- **Prolonged Mold Level Deviation**: >30 seconds outside normal range
- **Rapid Temperature Drop**: >50°C drop in 60 seconds
- **High Speed with Low Superheat**: Speed >1.5 m/min with superheat <20°C
- **Overall Defect Probability**: 15% base rate

#### Defect Labeling
- Binary classification (0 = no defect, 1 = defect)
- Deterministic rules mapping process excursions to defect probability
- Realistic correlation between process parameters and quality outcomes

### 3. Data Quality Features
- **Realistic Noise Patterns**: Gaussian noise with sensor-specific characteristics
- **Temporal Correlations**: Proper autocorrelation in sensor signals
- **Cross-Sensor Relationships**: Physical relationships between parameters
- **Missing Data Simulation**: Occasional sensor dropouts (< 1% of data points)

### configs/data_generation.yaml
```yaml
data_generation:
  num_casts: 1200
  cast_duration_minutes: 120
  sampling_rate_hz: 1
  random_seed: 42
  
  sensors:
    casting_speed:
      base_value: 1.2  # m/min
      noise_std: 0.05
      min_value: 0.8
      max_value: 1.8
    
    mold_temperature:
      base_value: 1520  # Celsius
      noise_std: 10
      min_value: 1480
      max_value: 1580
    
    # ... (additional sensor configurations)
  
  defect_simulation:
    defect_probability: 0.15
    defect_triggers:
      - prolonged_mold_level_deviation: 30  # seconds
      - rapid_temperature_drop: 50  # Celsius in 60 seconds
      - high_speed_with_low_superheat: true

  output:
    raw_data_format: "parquet"
    metadata_format: "json"
    train_test_split: 0.8
```

### Generated Data Structure
1. **Time Series Data**: 
   - DataFrame with columns: timestamp, casting_speed, mold_temperature, mold_level, cooling_water_flow, superheat
   - Index: datetime timestamps at 1Hz
   - ~7,200 rows per cast (120 minutes × 60 seconds)

2. **Metadata**:
   - Cast ID, generation timestamp, steel grade
   - Defect label (binary)
   - Process summary statistics
   - Defect trigger events (if any)

3. **File Organization**:
   ```
   data/
   ├── raw/
   │   ├── cast_timeseries_001.parquet
   │   ├── cast_timeseries_002.parquet
   │   └── ...
   ├── synthetic/
   │   ├── dataset_metadata.json
   │   └── generation_summary.json
   └── processed/
       └── (prepared for future feature engineering)
   ```

### scripts/generate_synthetic_data.py
Create executable script that:
- Loads configuration
- Initializes generator
- Runs full dataset generation
- Provides progress updates
- Saves generation summary

### Unit Tests (tests/test_data_generation.py)
- Test configuration loading
- Validate sensor value ranges
- Verify defect labeling logic
- Check output file formats
- Ensure reproducibility with random seed

### Integration Tests
- End-to-end dataset generation
- File output validation
- Data quality checks

## Success Criteria
- [ ] Generator produces 1200 synthetic casts
- [ ] Each cast has realistic 120-minute time series
- [ ] Defect rate approximately 15% as configured
- [ ] All sensor values within specified ranges
- [ ] Output files properly formatted and saved
- [ ] Random seed ensures reproducible results
- [ ] Unit tests pass with >90% coverage

## Dependencies on Other Components
- **Upstream**: Configuration management system
- **Downstream**: Feature engineering pipeline (Phase 2)
- **Related**: Data exploration notebooks

## Estimated Effort
- **Implementation**: 8-12 hours
- **Testing**: 4-6 hours
- **Documentation**: 2-3 hours

## Notes
- This synthetic generator will be replaced with real industrial data connectors in production
- Design for modularity to enable easy swapping of data sources
- Consider memory efficiency for large dataset generation
- Implement progress tracking for long-running generation jobs
